### PR TITLE
ExUnit: Show skipped tests as "S" and invalid in red in test output

### PR DESCRIPTION
**Pull request by wojtekmach**
Lately I noticed that I kept forgetting to re-enable tests that I previously marked as skipped. The total count and skipped count shows up at the bottom but I keep missing it and I wouldn't be surprised if other developers have the same problem. Here I propose a simple change that would make it obvious that a test was skipped. The "S" is stolen from minitest; a yellow "*" from RSpec or anything else, really, would work too. I also changed invalid to red.

Since it's a pretty trivial change I wasn't sure if it was worth discussing in elixir-lang-core, so I just went ahead and tried it out, but of course I can open up a thread.

Example run:
example

Example run with trace:
with trace

When some tests are skipped:
skipped

### CODE REVIEW

**Comment by josevalim** [comment]
Thank you!

We should not print skipped tests because the include/exclude mechanism is used also to filter tests that are never meant to run. For example, Ecto relies on that to control which tests to run for MySQL, PostgreSQL, etc. If you are filtering tests temporarily, consider doing it through the command line and using tags. Otherwise we would need to make a distinction between tests skipped locally and those via the command line.

**Comment by wojtekmach** [comment]
@josevalim thanks, that makes sense. I'll look into how ecto uses tags and see if there's a way to not compromise that use case with what I'm proposing.

> Otherwise we would need to make a distinction between tests skipped locally and those via the command line.

Is this something you're willing to consider in the future?

**Comment by josevalim** [comment]
If we have a clean solution it is something I would definitely consider.

**Comment by wojtekmach** [comment]
For anyone that finds this issue in the meantime - I was able to achieve what I proposed by doing the following in my project:

```elixir
# config/test.exs
use Mix.Config

defmodule CustomFormatter do
  defdelegate init(opts), to: ExUnit.CLIFormatter
  defdelegate trace_test_skip(test), to: ExUnit.CLIFormatter

  def handle_event({:test_finished, %ExUnit.Test{state: {:skip, _}} = test}, config) do
    if config.trace do
      IO.puts skipped(trace_test_skip(test), config)
    else
      IO.write skipped("S", config)
    end
    {:ok, %{config | tests_counter: config.tests_counter + 1,
                     skipped_counter: config.skipped_counter + 1}}
  end

  defdelegate handle_event(event, config), to: ExUnit.CLIFormatter

  defp colorize(escape, string, %{colors: colors}) do
    enabled = colors[:enabled]
    [IO.ANSI.format_fragment(escape, enabled),
     string,
     IO.ANSI.format_fragment(:reset, enabled)] |> IO.iodata_to_binary
  end

  defp skipped(msg, config) do
    colorize([:yellow], msg, config)
  end
end

config :ex_unit, formatters: [CustomFormatter]
```

**Comment by wojtekmach** [comment]
@josevalim FWIW seems Ecto is no longer using exclude/include for picking adapters to test, and instead modifies mix's test_path based on MIX_ENV ;-)

**Comment by josevalim** [comment]
No, we still rely heavily on it: https://github.com/elixir-lang/ecto/blob/master/integration_test/mysql/test_helper.exs#L5 :)

**Comment by wojtekmach** [comment]
doh, that's right. I didn't get the mysql test running locally yet, so I
haven't noticed. Thanks for the link!

2016-02-21 17:12 GMT+01:00 José Valim <notifications@github.com>:

> No, we still rely heavily on it:
> https://github.com/elixir-lang/ecto/blob/master/integration_test/mysql/test_helper.exs#L5
> :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/elixir/pull/4295#issuecomment-186851337>.
>



-- 
Wojtek Mach

**Comment by wojtekmach** [comment]
@josevalim sorry for still nagging about it :-)

TL;DR: how about allowing exunit to be configured with `show_skipped` option? (default value is up for debate, I'm leaning towards `true`)

long story:
I considered differentiating between "skipped" (as in `@tag :skip`) and "excluded" (with `exclude: ...` option) which would work well enough for Ecto (permanently disabled tests would be hidden, temporarily disabled tests would show up), but it would make the implementation way less elegant. You mentioned before about differentiating between command line arguments vs local settings, but that seems complicated too, and moreover, ideally, I'd prefer not to touch command line args pretty much never, and instead just do `mix test` majority of the time.

Allowing ExUnit to be configured with regards to `show_skipped` option could be IMHO good solution to the Ecto use case. I can think of 2 ways it could be configured (leaning towards option 2):

```elixir
# option 1
ExUnit.configure(show_skipped: true)
# option 2
ExUnit.configure(formatters: [{ExUnit.CLIFormatter, show_skipped: true}])
```

The downside is it'd introduce a little bit more complexity (and perhaps set a precedent for making it even more complicated in the future), configurability and things to remember. Perhaps the use case I'm presenting here isn't that common among devs. Other than the case I presented earlier, a potential upside for configurability for formatters is future extensibility: I can easily imagine doing something like following in the future: `formatters: [CLIFormatter, {HTMLFormatter, output: "tmp/test_html"}]`

**Comment by josevalim** [comment]
I am :-1:. I would rather introduce a new state, if justified, than add more configuration options. Basically, you are not using skipped for what it was designed to do, and I would rather introduce something proper than try to make it work like it should not.
